src/PackagePrivate/DatabaseTermIdsAcquirer.php

```diff
@@ -0,0 +1,420 @@
+<?php
+
+namespace Wikibase\TermStore\MediaWiki\PackagePrivate;
+
+use Wikimedia\Rdbms\IDatabase;
+
+class DatabaseTermIdsAcquirer implements TermIdsAcquirer {
+
+	const TABLE_TEXT = 'wbt_text';
+	const TABLE_TEXT_IN_LANG = 'wbt_text_in_lang';
+	const TABLE_TERM_IN_LANG = 'wbt_term_in_lang';
+
+	/**
+	 * @var Database $dbw
+	 */
+	private $dbw;
+
+	/**
+	 * @var Database $dbr
+	 */
+	private $dbr;
+
+	public function __construct(
+		IDatabase $dbw,
+		IDatabase $dbr,
+		TypeIdsAcquirer $typeIdsAcquirer
+	) {
+		$this->dbw = $dbw;
+		$this->dbr = $dbr;
+		$this->typeIdsAcquirer = $typeIdsAcquirer;
+	}
+
+	public function acquireTermIds( array $termsArray ): array {
+		$termsArray = $this->mapToTextIds( $termsArray );
+		$termsArray = $this->mapToTextInLangIds( $termsArray );
+		$termsArray = $this->mapToTypeIds( $termsArray );
+		return $this->mapToTermInLangIds( $termsArray );
+	}
+
+	/**
+	 * replace root keys containing type names in termsArray
+	 * with their respective ids in wbt_type table
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type1' => [ ... ],
+	 *		'type2' => [ ... ],
+	 *		...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		<typeId1> => [ ... ],
+	 *		<typeId2> => [ ... ],
+	 *		...
+	 *	]
+	 */
+	private function mapToTypeIds( array $termsArray ) {
+		$typeIds = $this->typeIdsAcquirer->acquireTypeIds( array_keys( $termsArray ) );
+
+		$termsArrayByTypeId = [];
+		foreach ( $typeIds as $type => $typeId ) {
+			$termsArrayByTypeId[$typeId] = $termsArray[$type];
+		}
+		return $termsArrayByTypeId;
+	}
+
+	/**
+	 * replace text at termsArray leaves with their ids in wbt_text table
+	 * and return resulting array
+	 *
+	 * @param array $termsArray terms per type per language:
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => 'term' | [ 'term1', 'term2', ... ] ], ...
+	 *		], ...
+	 *	]
+	 *
+	 * @return array
+	 *	[
+	 *		'type' => [
+	 *			[ 'language' => [ <textId1>, <textId2>, ... ] ], ...
+	 *		], ...
+	 *	]
+	 */
+	private function mapToTextIds( array $termsArray ) {
+		$texts = [];
+
+		array_walk_recursive( $termsArray, function ( $text ) use ( &$texts ) {
+			$texts[] = $text;
+		} );
+
+		$textIds = $this->acquireTextIds( array_unique( $texts ) );
```

Comment:
`array_unique` preserves array keys, so the resulting array may have “gaps” in the keys. It’s probably best to call `array_values` on the result to turn it into a plain list.

Reply:
this doesn't matter in my case here .. I do not rely on keys in that array

Reply:
You were so right. I had to use `array_values` to reset the keys here since `IDatabase::insert` implementation did rely on them starting from 0.

Reply:
Well I had to use it somewhere else, not in this case specifically .. but I didn't expect that to be necessary